Documentation for elasticsearch reindex endpoints (#262)
nasa/cumulus-api: content/elasticsearch.md

## Elasticsearch

These endpoints provide an interface into the Elasticsearch SDK to perform functions like Elasticsearch reindex.

Cumulus uses [index aliases](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html) to allow reindexing with no downtime.

## Reindex

> The reindex command creates a new index and reindexes the source index to the new, destination index. This uses the [Elasticsearch reindex functionality](https://www.elastic.co/guide/en/elasticsearch/reference/5.5/docs-reindex.html), but includes setting up the new index.

Reviewer: Not sure if it matters, but by default people will get ES 5.3 deployed, so should we point to that version of the ES docs?

> Note that there will now be two copies of your index. Your Cumulus instance will still point to the old copy until you perform the `change-reindex` function described below.

Reviewer: ```suggestion
Note that there will now be two copies of your index. Your Cumulus instance will still point to the old copy until you perform the `change-index` function described below.
```

> An alias should not be specified unless you have a specific alias configured. If a source index is not specified, it will default to the index from the alias. If you want to name the destination index something particular, you can specify a name, otherwise the destination index name will default to 'cumulus-year-month-day' with today's date. If you specify a destination index name, it must be an index that does not already exist in your cluster.

Reviewer: ```suggestion
An alias should not be specified unless you have a specific alias configured. If a source index is not specified, it will default to the index from the alias. If you want to name the destination index something particular, you can specify a name, otherwise the destination index name will default to 'cumulus-year-month-day' with today's date (e.g. `cumulus-4-12-2019`). If you specify a destination index name, it must be an index that does not already exist in your cluster.
```

> An alias should not be specified unless you have a specific alias configured. If a source index is not specified, it will default to the index from the alias. If you want to name the destination index something particular, you can specify a name, otherwise the destination index name will default to 'cumulus-year-month-day' with today's date. If you specify a destination index name, it must be an index that does not already exist in your cluster.

Reviewer: Re-reading what the reindex endpoint does, isn't the expectation of an alias strictly enforced? Because if you don't specify a source index, it tries to look it up from the alias, but if you do specify one, it checks to make sure that source index exists for the alias?

https://github.com/nasa/cumulus/pull/931/files#diff-536815ac2c6bf81fbd2e9d46e354b1f4R73

Reviewer: Ah, that is fine because there will always be an alias, but by default it will be the one Cumulus deploys with. So we're just saying don't specify a custom alias unless you created one.

Since reindex can be a long running operation, the `reindex-status` endpoint must be queried to see if the operation has completed.

#### Example requests

```curl
$ curl --request POST https://example.com/v1/elasticsearch/reindex --header 'Authorization: Bearer ReplaceWithTheToken'
```

```curl
$ curl --request POST https://example.com/v1/elasticsearch/reindex --header 'Authorization: Bearer ReplaceWithTheToken' --data '{
  "aliasName": "cumulus-alias",
  "sourceIndex": "cumulus-4-4-2019",
  "destIndex": "cumulus-new-index"
}'
```

#### Example response

```json

> { "message": "Reindexing from cumulus-4-4-2019 to cumulus-4-12-2019. Check the reindex-status endpoint for status.}

Reviewer: Wouldn't the response be `Reindexing from cumulus-4-4-2019 to cumulus-new-index`?

Author: There are two request examples in there and I wanted to show how it would generate a timestamped index. I switched the format though so each example request has its own example response.

                    "docs": {
                        "count": 1,
                        "deleted": 0
                    }
                },
                "total": {
                    "docs": {
                        "count": 1,
                        "deleted": 0
                    }
                }
            },
            "cumulus-2019-4-12": {
                "primaries": {
                    "docs": {
                        "count": 25569,
                        "deleted": 0
                    }
                },
                "total": {
                    "docs": {
                        "count": 25569,
                        "deleted": 0
                    }
                }
            },
            "cumulus-2019-4-9": {
                "primaries": {
                    "docs": {
                        "count": 25640,
                        "deleted": 0
                    }
                },
                "total": {
                    "docs": {
                        "count": 25640,
                        "deleted": 0
                    }
                }
            }
        }
    }
}
```

## Change Index

> Change index switches Cumulus to point to the new destination index, rather than the source index. You may choose to delete your source index during this operation using the `deleteSource` parameter.

Reviewer: ```suggestion
Change index switches the Elasticsearch alias to point to the new destination index, rather than the source index. You may choose to delete your source index during this operation using the `deleteSource` parameter.
```

`sourceIndex` and `destIndex` are required parameters.

In the context of reindex, you'd call `change-index`, following reindex completion to start using the new index with no downtime.

#### Example request

```curl
$ curl --request POST https://example.com/v1/elasticsearch/change-index --header 'Authorization: Bearer ReplaceWithTheToken' --data '{
  "aliasName": "cumulus-alias",
  "sourceIndex": "cumulus-4-4-2019",
  "destIndex": "cumulus-new-index",
  "deleteSource": false
}'
```

#### Example response

```json

> { "message": "Reindex success - alias cumulus now pointing to cumulus-4-4-2019" }

Reviewer: Wouldn't this now be pointing to `cumulus-new-index`?
